packages/astro/src/core/create-vite.ts

```diff
@@ -128,9 +128,26 @@ export async function createVite(
 	let result = commonConfig;
 	result = vite.mergeConfig(result, astroConfig.vite || {});
 	result = vite.mergeConfig(result, commandConfig);
+	sortPlugins(result);
 	return result;
 }
 
+function getPluginName(plugin: vite.PluginOption) {
+	if (plugin && typeof plugin === 'object' && !Array.isArray(plugin)) {
+		return plugin.name;
+	}
+}
+
+function sortPlugins(result: ViteConfigWithSSR) {
+	// HACK: move mdxPlugin to top because it needs to run before internal JSX plugin
+	const mdxPluginIndex = result.plugins?.findIndex(plugin => getPluginName(plugin) === '@mdx-js/rollup') ?? -1;
+	if (mdxPluginIndex === -1) return;
+	const jsxPluginIndex = result.plugins?.findIndex(plugin => getPluginName(plugin) === 'astro:jsx') ?? -1;
+	const mdxPlugin = result.plugins?.[mdxPluginIndex];
+	result.plugins?.splice(mdxPluginIndex, 1);
+	result.plugins?.splice(jsxPluginIndex, 0, mdxPlugin);
+}
```

Comment:
Not sure if there's a cleaner way to do this?

Reply:
I think a `sort()` would work here as well, but not sure if it'd be any cleaner at the end of the day 🤷

Reply:
I think we could explore our own `enforce` hook eventually, but that's out of scope for this PR. The built-in one is not granular enough. 

I think `sort()` could work but I couldn't come up with something that's less verbose than this.

If anyone has a concrete suggestion I'll take it, otherwise I'm okay with this approach for now.

Reply:
👍 Yeah I started on a `sort()` only to find it was just as complex, the `splice` approach looks good to me